WIP not for review - Add IANA language subtag registry support

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Uploaded WIP branch to check build and for AI feedback on what I've implemented so far for the BCP 47 IANA registry, registry file and accessor classes.

This is only part of the work for BCP 47, there is another WIP branch for the tag parsing. Some validation is done, normalization is very approximative and non-compliant at this time.

After those yet another branch will be needed to integrate all that with stream selection / UI.

CrystalP committed:
[i18n] Create sub tag registry, related classes and tests

CrystalP committed:
[SmartPlaylist] Move RFC3339 full-date date validation into CDateTime

CrystalP commented:
@claude please review again

CrystalP commented:
@claude critical issue 1 is wrong. value.m_description is const so the copy is necessary and the suggestions std::string& desc and auto& desc don't compile. Ignore the issue in the PR from now on. Critical issue 2 is known and won't be addressed in this PR. Ignore the issue in the PR from now on.

CrystalP commented:
@garbear could you please start copilot again.
The claude reached its limit, build should be fixed. It looks like FreeBSD is unable to properly validate dates, causing an incorrect unit test failure. 2025-30-42 is definitely not in my calendar :-)

CrystalP commented on xbmc/utils/i18n/SubTagRegistry.h:
```diff
@@ -0,0 +1,136 @@
+};
+
+XBMC_GLOBAL_REF(CSubTagRegistry, g_subTagRegistry);
+#define g_subTagRegistry XBMC_GLOBAL_USE(CSubTagRegistry)
```
You're probably right. One global to rule them all :-)

garbear commented:
> there is 40000 lines of text registry to be maintained i think this is a bit too much.... If exists a library would be preferable

Fair point, that's a heavy burden, tho data is less burdensome than code. And yes, if an existing dependencies can save code, I'm all for it. Not commenting as an expert of ICU but as a voice that a nonzero maintenance burden can still be a net positive.

fuzzard commented:
Presumably the file is just the data ripped from https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry
If so, that would be considered maintainable (single location, date verifiable). Just would need some sort of documentation letting people know that is what is used to update/maintain it.

CastagnaIT commented:
have you tried to see if ICU library can save us to maintain all this?

CastagnaIT commented:
If so it seems easy, yeah it would be appropriate to document clearly in code how to perform the updates
However i also wonder how much parsing that large file affects slow devices

CrystalP committed:
[i18n] Launch the registry with ServiceManager

CrystalP commented:
rebased.

CrystalP committed:
[i18n] Fix BCP47 parsing regexp to allow 5*8alphanum variants

CrystalP committed:
[i18n] Add BCP47 registry support to LangCodeExpander ReverseLookup / adapt Lookup

CrystalP committed:
[i18n] Inject registry into BCP47 tag for parsing/validation

CrystalP committed:
[i18n] Refactor BCP47 Tag tests

CrystalP committed:
[i18n] Make the BCP47 formatter a friend and remove field accessors

CrystalP committed:
[i18n] Rename FormatXXX functions of BCP47 formatter to AppendXXX

CrystalP committed:
[i18n] Move the tag debug print format into the formatter

CrystalP committed:
[i18n] Format Bcp47 using the registry

CrystalP committed:
[i18n] Remove unnecessary BCP47 default comparison operator

CrystalP commented:
rebased, added some commits to take advantage of the registry when parsing (nfo & scraper import) and formatting the tags for output (ex. language infolabels)
fixed a few bugs in the bcp47 classes merged earlier.
todo:
decide whether to do something about startup performance or not - need feedback on slow device. maybe later.
add a bit of DI for better testing
should probably remove fallback to iso 639/3166 data in the formatter

CrystalP committed:
[Playlists] Add Original Language Smart Playlist Rule.

CrystalP committed:
[playlists] Add some validation for audio/subtitle/original language

CrystalP committed:
[guilib] Add formatting function on GUIEditControl lost focus

CrystalP committed:
[playlists] Reformat the original language after editing a rule

CrystalP committed:
[playlists] Move smart playlist validators into a dedicated .cpp

CrystalP committed:
[i18n] Format tags with the shortest available subtag descriptions

CrystalP commented:
- Implemented the use of the shortest of the available descriptions
- Added a couple commits implementing a smartplaylist rule for the original language (will be a separate PR)

The description of ao1990 is actually worse because it's split on two lines in the file but is put back together by the registry reading code, making it 101 characters.
However commonly used tags only include the language and the country and/or script at best and a cursory look didn't show overly long descriptions.

As you wrote they can be overridden with advanced settings or language addons.

Still, I implemented the idea of the ellipsis to limit description length to 30 characters.
It's not great for region 419 `Latin America and the Caribbean`, truncated to `Latin America and the Carib...`

Another thing, there could be up to 3 extlang or multiple variant subtags in a tag. Limit the length of each extlang/variant or the length of the group after concatenation of the individual subtags?

CrystalP committed:
[i18n] Share in-memory registry source to improve unit testing of Bcp47

CrystalP committed:
[i18n] Limit English tag display to 30 characters + BCP47 when over 30

CrystalP commented:
I think it's ready to go - there will be time to finetune before release.
Closing in favor of PR #27893, let's continue discussing there.
